src/main/scala/org/broadinstitute/dsde/workbench/sam/service/ResourceService.scala

```diff
@@ -17,86 +15,110 @@ import scala.concurrent.{ExecutionContext, Future}
+      _ <- Future.traverse(policiesToDelete){accessPolicyDAO.deletePolicy}
+    //      _ <- accessPolicyDAO.deleteResource(resource) //todo: why does it work even if we don't delete this level?
+    } yield ()
+  }
+
+  def hasPermission(resource: Resource, action: ResourceAction, userInfo: UserInfo): Future[Boolean] = {
+    listUserResourceActions(resource, userInfo).map { _.contains(action) }
+  }
+
+  def listUserResourceActions(resource: Resource, userInfo: UserInfo): Future[Set[ResourceAction]] = {
+    listResourceAccessPoliciesForUser(resource, userInfo).map { matchingPolicies =>
       matchingPolicies.flatMap(_.actions)
     }
   }
 
-  def listUserResourceRoles(resourceTypeName: ResourceTypeName, resourceName: ResourceName, userInfo: UserInfo): Future[Set[ResourceRoleName]] = {
-    listResourceAccessPoliciesForUser(resourceTypeName, resourceName, userInfo).map { matchingPolicies =>
-      matchingPolicies.flatMap(_.role)
+  def listUserResourceRoles(resource: Resource, userInfo: UserInfo): Future[Set[ResourceRoleName]] = {
+    listResourceAccessPoliciesForUser(resource, userInfo).map { matchingPolicies =>
+      matchingPolicies.flatMap(_.roles)
     }
   }
 
-  private def listResourceAccessPoliciesForUser(resourceTypeName: ResourceTypeName, resourceName: ResourceName, userInfo: UserInfo): Future[Set[AccessPolicy]] = {
-    for {
-      policies <- accessPolicyDAO.listAccessPolicies(resourceTypeName, resourceName)
-      groups <- directoryDAO.listUsersGroups(userInfo.userId)
-    } yield {
-      policies.filter { policy =>
-        policy.subject match {
-          case user: WorkbenchUserId => userInfo.userId == user
-          case group: WorkbenchGroupName => groups.contains(group)
+  private def listResourceAccessPoliciesForUser(resource: Resource, userInfo: UserInfo): Future[Set[AccessPolicy]] = {
+    accessPolicyDAO.listAccessPoliciesForUser(resource, userInfo.userId)
+  }
+
+  //Overwrites an existing policy (keyed by resourceType/resourceId/policyName), saves a new one if it doesn't exist yet
+  def overwritePolicy(policyName: String, resource: Resource, policyMembership: AccessPolicyMembership, userInfo: UserInfo): Future[AccessPolicy] = {
+    requireAction(resource, ResourceAction("altermembers"), userInfo) {
+      val subjectsFromEmails = Future.traverse(policyMembership.memberEmails) {
+        directoryDAO.loadSubjectFromEmail
+      }.map(_.flatten)
+
+      val email = s"policy-${resource.resourceTypeName.value}-${resource.resourceId.value}-${policyName}@dev.test.firecloud.org" //TODO: Make sure this is a good/unique naming convention and keep Google length limits in mind
+
+      subjectsFromEmails.flatMap { members =>
+        val newPolicy = AccessPolicy(policyName, resource, WorkbenchGroup(WorkbenchGroupName(policyName), members, WorkbenchGroupEmail(email)), policyMembership.roles, policyMembership.actions)
+
+        accessPolicyDAO.listAccessPolicies(resource).flatMap { policies =>
```

Comment:
should probably have a loadAccessPolicy function

Reply:
instead of listing all policies for a resource, implement a function in the dao to load only the named policy